src/charm.py

```diff
@@ -10,41 +10,68 @@
 import os
 from charms.operator_libs_linux.v2 import snap  # type: ignore
 from ops.model import ActiveStatus, MaintenanceStatus
-from snap_management import SnapSpecError, SnapInstallError, SnapServiceError, install_snap, node_exporter_snap_name, opentelemetry_collector_snap_name
+from snap_management import (
+    SnapSpecError,
+    SnapInstallError,
+    SnapServiceError,
+    install_snap,
+    node_exporter_snap_name,
+    opentelemetry_collector_snap_name,
+)
+from singleton_snap import SingletonSnapManager
 
 # Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
 
 SNAPS = [opentelemetry_collector_snap_name, node_exporter_snap_name]
 
+
 class OpentelemetryCollectorOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
         self._reconcile()
 
-
     def _reconcile(self):
         self._install()
-        self._remove()
+        self._stop()
         self.unit.status = ActiveStatus()
 
     def _install(self) -> None:
         if self.hook != "install":
             return
 
+        manager = SingletonSnapManager(self.unit.name)
         for snap_package in SNAPS:
-            self._install_snap(snap_package)
-            self._start_snap(snap_package)
-
-    def _remove(self) -> None:
-        if self.hook != "remove":
+            manager.register(snap_package)
+            with manager.snap_operation(snap_package):
+                self._install_snap(snap_package)
```

Comment:
We will need to use `get_revisions()` somewhere, and include logic such as:

```
if our_revision > max(get_revisions()):
    self._install_snap()
```

Reply:
Done here: [`87c9988` (#6)](https://github.com/canonical/opentelemetry-collector-operator/pull/6/commits/87c99886dd3b1da510a4cba4919b62c8a2751dd7)